blog/content/post/2020-11-30-coreboot_for_power.md

```diff
@@ -0,0 +1,154 @@
+
+#### For users
+
+Boot time reduction was one of the main reasons for doing this port. Hostboot
+itself takes more than 1 minute to run. This is mostly spent on accessing flash
+memory. It is split into multiple PNOR (Processor NOR, POWER name for flash)
+partitions, summing up to just below 32 megabytes. It runs from cache until it
+trains main memory, so it is limited to just 10 MB (size of L3 cache) split
+between two cores. To make it possible to load and execute that amount of code,
+[Hostboot uses on-demand paging](https://youtu.be/fTLsS_QZ8us?t=1559). It is
+similar to swap used by operating systems, except that it uses ROM instead of
+writable media, so only code and constant data can be discarded from memory (L3
+cache in this case).
+
+Another issue was nicely formulated by Timothy Pearson from Raptor CS:
+
+> Hostboot is a VM to run FSI routines, and the FSI routines are written one by
+> one by the hardware engineers designing the silicon itself.  This leads to
+> overly complex and difficult to understand code, as you can see.
+
+FSI is one of the buses Hostboot has to initialise. By simplifying the code we
+can make it **easier to understand** (no need to translate virtual addresses to
+the physical ones), **faster** (by discarding a dozen of function calls before
+the proper write to the register happens) and **smaller**. Smaller code means
+less time spent on reading from flash, but also more space for other components,
+like new drivers for Skiroot.
+
+#### For the coreboot community
+
+coreboot would get a **new architecture** supported. Although there are stubs,
+cross-compiler and libraries for PPC, the coreboot itself isn't in working state
+when it comes to this architecture as of now. It would also get a **new, truly
+open** ([unless you want SATA controller](https://wiki.raptorcs.com/wiki/PM8068))
+platform.
+
+PowerPC by default uses big endian. It is configurable at runtime, but the
+libraries included in coreboot-sdk are compiled for BE. This gives a perfect
+opportunity to validate that **coreboot works also for big endian platforms**.
+By now we confirmed that there are parts in CBFS and FMAP which need small
+fixes. The endianness of fields also needs to be properly documented for those
+structures, as they can be used both by a platform it is started on, as well as
+an application on a platform it is built on. Those two do not have to use the
+same endianness.
+
+The biggest achievement of this port would undoubtedly be bringing support for
+**native DDR4 initialisation and training**. Even though the details of
+accessing the hardware are platform-specific, the general process stays the
+same. One caveat might be that Talos II supports only RDIMM (registered, or
+buffered DIMM) and that is the only kind of memory sticks we are able to test,
+however even if it won't work out of the box for different DIMM flavours, it
+should give a good base to build on.
```

Comment:
give to whom? maybe it should be a good base?

Reply:
@krystian-hebel benefit of that could be way better explained TBH this is a huge deal. I'm not sure what platforms operate on DDR with RDIMM, but we can imagine the situation that we get rid of FSP-M part or piece or PSP :thinking: and getting rid of memory initialization. Even if not x86 then RISC-V would benefit.

Reply:
@KBanecka developers (newer DDR standards documentation without blobs or NDA, first, then end-users after base is worked on and others jump in code base.